python/ray/data/_internal/arrow_block.py

```diff
@@ -111,6 +111,23 @@ def __init__(self):
             raise ImportError("Run `pip install pyarrow` for Arrow support")
         super().__init__(pyarrow.Table)
 
+    def add_block(self, block: Any) -> None:
+        if not isinstance(block, pyarrow.Table) and not isinstance(block, bytes):
+            raise TypeError(
+                f"Got a block of type {type(block)}, expected {self._block_type}."
+                "If you are mapping a function, ensure it returns an "
+                "object with the expected type. Block:\n"
+                f"{block}"
+            )
+        accessor = BlockAccessor.for_block(block)
+        if isinstance(block, bytes):
+            reader = pyarrow.ipc.open_stream(block)
+            self._tables.append(reader.read_all())
+        else:
+            self._tables.append(block)
+        self._tables_size_bytes += accessor.size_bytes()
+        self._num_rows += accessor.num_rows()
```

Comment:
@kira-lin Why is this necessary? The existing [`TableBlockBuilder.add_block()` logic](https://github.com/ray-project/ray/blob/ff66e0ef1bc248a547a88f3eb96512d986b9ed57/python/ray/data/_internal/table_block.py#L77-L87) should convert the bytes into an Arrow table: (delegation)[https://github.com/ray-project/ray/blob/ff66e0ef1bc248a547a88f3eb96512d986b9ed57/python/ray/data/block.py#L394-L397], (conversion)[https://github.com/ray-project/ray/blob/ff66e0ef1bc248a547a88f3eb96512d986b9ed57/python/ray/data/_internal/arrow_block.py#L149-L152]

Is this an issue with the instance check in `TableBlockBuilder.add_block()`? If so, we can change the block type supplied by the `ArrowBlockBuilder` constructor to its parent's constructor to a `(pyarrow.Table, bytes)` tuple, since `isinstance()` checks work with a tuple of types, and `TableBlockBuilder._block_type` is only used for that instance check.

Reply:
oh, I see. I tried Union, didn't work. but this alone is not enough to solve the problem. Please see the update.